inject/src/main/java/io/micronaut/context/DefaultApplicationContext.java

```diff
@@ -376,223 +369,154 @@ private static String resolveEachPropertyMissingBeanMessage(Qualifier<?> qualifi
     }
 
     @Override
-    protected <T> Collection<BeanDefinition<T>> transformIterables(BeanResolutionContext resolutionContext, Collection<BeanDefinition<T>> candidates, boolean filterProxied) {
-        if (!candidates.isEmpty()) {
-
-            List<BeanDefinition<T>> transformedCandidates = new ArrayList<>();
-            for (BeanDefinition<T> candidate : candidates) {
-                if (candidate.isIterable()) {
-                    if (candidate.hasDeclaredStereotype(EachProperty.class)) {
-                        transformEachPropertyBeanDefinition(resolutionContext, candidate, transformedCandidates);
-                    } else if (candidate.hasDeclaredStereotype(EachBean.class)) {
-                        transformEachBeanBeanDefinition(resolutionContext, candidate, transformedCandidates, filterProxied);
-                    }
-                } else if (candidate.hasStereotype(ConfigurationReader.class)) {
-                    transformConfigurationReaderBeanDefinition(resolutionContext, candidate, transformedCandidates);
-                } else {
-                    transformedCandidates.add(candidate);
-                }
-            }
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Finalized bean definitions candidates: {}", candidates);
-                for (BeanDefinition<?> definition : transformedCandidates) {
-                    LOG.debug("  {} {} {}", definition.getBeanType(), definition.getDeclaredQualifier(), definition);
-                }
+    protected <T> void collectIterableBeans(BeanResolutionContext resolutionContext, BeanDefinition<T> iterableBean, Set<BeanDefinition<T>> targetSet) {
+        try(BeanResolutionContext rc = newResolutionContext(iterableBean, resolutionContext)) {
+            if (iterableBean.hasDeclaredStereotype(EachProperty.class)) {
```

Comment:
You can add `if (candidate.isIterable())`, which should be a compile-time evaluated value.